### PR TITLE
Adding supervised data config

### PR DESCRIPTION
**Pull request by TheQuantumFractal**

### CODE REVIEW

File: src/levanter/data/text.py [hunk 1]
```diff
@@ -584,6 +588,103 @@ def tagged_eval_sets(
...
+
+    cache_dir: str = "cache/"
+
+    tags: Optional[List[str]] = None
+    """tags for the dataset. Typically the name of the dataset in the config will be added as a tag as well"""
+    name: Optional[str] = None  # name for hf dataset
+
+    validation_urls: List[str] = ()  # type:ignore
+
+    # def token_seq_dataset(
```
Comment (ahmeda14960) [inline src/levanter/data/text.py:603]: Can you remove this code if it's commented out?

File: src/levanter/main/train_lm.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from levanter import callbacks
 from levanter.checkpoint import load_checkpoint
 from levanter.compat.hf_checkpoints import HFCompatConfig, save_hf_checkpoint_callback
-from levanter.data.text import CausalLmDataset, LMDatasetConfig, LMMixtureDatasetConfig
+from levanter.data.text import CausalLmDataset, LMDatasetConfig, LMMixtureDatasetConfig, LMSupervisedDatasetConfig
```
Comment (ahmeda14960) [inline src/levanter/main/train_lm.py:19]: Should you maybe make a different file? I think train_lm is typically expected within levanter to be reserved for pre-training and is specific to the autoregressive loss. I think there's separate files for fine-tuning alpaca for example. 

I'm not sure what the convention is cc: @dlwh

File: src/levanter/data/text.py [hunk 3]
```diff
@@ -584,6 +588,103 @@ def tagged_eval_sets(
...
+    """
+    Prepare an example for training. This function converts the (cached) batch encoding into an LmExample.
+
+    It goes through the following steps:
+
+    1. Pad the batch to the maximum length.
+    2. Mask out the input and prompt if requested.
+    3. Create an LmExample with the input_ids as the input and the next token as the target.
+    """
+    with local_cpu_mesh():
```
Comment (ahmeda14960) [inline src/levanter/data/text.py:655]: Why with local cpu mesh here? how would this work if you wanted to do training?

File: src/levanter/data/text.py [hunk 4]
```diff
@@ -584,6 +588,103 @@ def tagged_eval_sets(
...
+
+def mk_supervised_dataset(config: LMSupervisedDatasetConfig, tokenizer: PreTrainedTokenizerBase):
+    import levanter.data
+
+    validation_urls = [url for url_pat in config.validation_urls for url in fsspec_expand_glob(url_pat)]
+    dataset = levanter.data.datasource_from_jsonl(validation_urls)
+
+    output_exemplar = {"input_ids": np.zeros((0,), dtype=np.int32), "sources_len": np.zeros((), dtype=np.int32)}
+
+    dataset = dataset.map_batches(lambda ex: preprocess_supervised_example(ex, tokenizer), batch_size=128, num_cpus=num_cpus_used_by_tokenizer(tokenizer), output_exemplar=output_exemplar)  # type: ignore
```
Comment (ahmeda14960) [inline src/levanter/data/text.py:680]: How exactly is ```output_exemplar``` used ? I see that we modified the datasets to accept it as an argument but I don't see any changed logic that actually uses it.

File: src/levanter/data/text.py (hunk 1 above)
Comment (TheQuantumFractal) [reply src/levanter/data/text.py:603]: It was throwing an error from the precommit. I need to check with David if we need these functions. If so, I'll need to fix some of these

File: src/levanter/data/text.py (hunk 3 above)
Comment (TheQuantumFractal) [reply src/levanter/data/text.py:655]: This is just for preprocessing the supervised examples on the CPU (tokenizing and masking them). They are put on TPUs after.

File: src/levanter/data/text.py (hunk 4 above)
Comment (TheQuantumFractal) [reply src/levanter/data/text.py:680]: it's used when building the cache. Basically the rank of the input ids and sources len has to be known in order to load from or save to the cache (and the original sharded datasource struct doesn't support this)

File: src/levanter/main/train_lm.py (hunk 2 above)
Comment (TheQuantumFractal) [reply src/levanter/main/train_lm.py:19]: For supervised finetuning, we maybe should make a separate file but the changes in this file are for measuring eval loss on a supervised task.

Comment (ahmeda14960) [reply src/levanter/main/train_lm.py:19]: Got it LGTM

File: src/levanter/data/text.py [hunk 5]
```diff
@@ -584,6 +588,103 @@ def tagged_eval_sets(
...
+    #         return {}
+
+    # Add tagged eval set with split for auxiliary and validation dataset
+
+
+def preprocess_supervised_example(batch, tokenizer: PreTrainedTokenizerBase):
+    sources = [example["input"] for example in batch]
+
+    targets = [f"{example['output']}" for example in batch]
+    # TODO: this seems pretty wasteful since you end up tokenizing twice, but it's how the original code does it.
```
Comment (dlwh) [inline src/levanter/data/text.py:632]:
```suggestion
    # TODO: this seems pretty wasteful since you end up tokenizing twice, but it's how alpaca does it.
```

File: src/levanter/data/text.py (hunk 1 above)
Comment (dlwh) [reply src/levanter/data/text.py:603]: rm and don't have LMSupervisedDatasetConfig implement LMDatasetSourceConfig

File: src/levanter/data/text.py [hunk 6]
```diff
@@ -390,6 +390,22 @@ def batch_size(self) -> int:
         return self._batch_size
 
 
+def fsspec_expand_glob(url):
```
Comment (dlwh) [inline src/levanter/data/text.py:393]: can you move to utils/fsspec_utils

File: src/levanter/data/text.py (hunk 3 above)
Comment (dlwh) [reply src/levanter/data/text.py:655]: Yeah it has to be local_cpu_mesh. I'll try to make that part get hidden

**Review by dlwh — APPROVED** [review]
Thanks!
